### PR TITLE
Enable FL Server in SGX

### PR DESCRIPTION
**Pull request by Hanyu-Jin**
## Description

Enable FL Server in SGX.

### 1. Why the change?

As FL server has the data from several clients, we use SGX to make sure that the process of data calculation on FL server is safe.

### 2. User API changes

Users will need to create a new container, create a script that invoke start_fl_server.py and run it to run FL Server in SGX.
Use -c/--client-num= to set the number of client.
Use -p/--port= to set port.

### 3. Summary of the change 

Pack relative codes of FL server.
Create start_fl_server.py to call FL server.

### 4. How to test?
- [ ] Unit test
- [ ] GitHub Action

### 5. New dependencies

- [ ] New Python dependencies
       - grpcio 1.47.0
       - protobuf 3.19.4

### CODE REVIEW

**Comment by Litchilitchy** [comment]
Some updates should not belong to this PR (maybe you pulled some code and did not sync with master). Please check you have merge/rebase to the master

**Comment by Litchilitchy** [comment]
This pr also includes the changes in #5277

**Comment by Litchilitchy** [comment]
Does `grpc` package already include `protobuf`?

**Comment by Hanyu-Jin** [comment]
#

**Comment by Hanyu-Jin** [comment]
The edition of grpcio and protobuf is compatible. The environment contains both of them.

**Review by Litchilitchy — APPROVED** [review]

File: python/ppml/example/pytorch_nn_lr/pytorch-nn-lr-tutorial.md [hunk 1]
```diff
@@ -96,19 +96,30 @@ result = ppl.predict(x)
...
+You can set port with `-p` or `--port` and set client number with `-c` or `--client-num` while the default settings are `port=8980` and `client-num=2`.
+Then run the script,
+```bash
+bash start-python-fl-server-sgx.sh
 ```
-Then start the FL Server
+or run it outside of the container.
 ```bash
-python BigDL/python/ppml/src/bigdl/ppml/fl/nn/fl_server.py
+docker exec -it YOUR_CONTAINER bash start-python-fl-server-sgx.sh
```
Comment (Litchilitchy) [inline python/ppml/example/pytorch_nn_lr/pytorch-nn-lr-tutorial.md:122]: This script does not exist in this PR. Did you put it in docker image?

Comment (Hanyu-Jin) [reply python/ppml/example/pytorch_nn_lr/pytorch-nn-lr-tutorial.md:122]: The origin name of script in spark-k8s-client-attestation-devel-HY is start-python-flserver2-sgx.sh.
I've changed the name now

File: scala/ppml/src/assembly/ppml-assembly.xml [hunk 2]
```diff
@@ -0,0 +1,26 @@
+<assembly xmlns="http://maven.apache.org/ASSEMBLY/2.0.0"
```
Comment (glorysdj) [inline scala/ppml/src/assembly/ppml-assembly.xml:1]: this is not needed

File: ppml/trusted-big-data-ml/python/docker-graphene/Dockerfile [hunk 3]
```diff
@@ -277,6 +277,7 @@ ADD ./test-suites /ppml/trusted-big-data-ml/work/test-suites
 ADD ./_dill.py.patch ./_dill.py.patch
 ADD ./python-uuid.patch ./python-uuid.patch
 ADD ./python-pslinux.patch ./python-pslinux.patch
+ADD ../../../../python/ppml/scripts/start-fl-server.py /ppml/trusted-big-data-ml/work/examples
```
Comment (glorysdj) [inline ppml/trusted-big-data-ml/python/docker-graphene/Dockerfile:280]: ADD with github raw url

Comment (glorysdj) [reply ppml/trusted-big-data-ml/python/docker-graphene/Dockerfile:280]: or prepare it in # stage.3 bigdl

File: ppml/trusted-big-data-ml/python/docker-graphene/start-scripts/start-python-fl-server-sgx.sh [hunk 4]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+cd /ppml/trusted-big-data-ml
+/graphene/Tools/argv_serializer bash -c " /opt/jdk8/bin/java\
+        -cp '/ppml/trusted-big-data-ml/work/spark-3.1.2/conf/:/ppml/trusted-big-data-ml/work/spark-3.1.2/jars/*'\
+        -Xmx10g org.apache.spark.deploy.SparkSubmit\
+        --master 'local[4]'\
+        /ppml/trusted-big-data-ml/work/examples/start-fl-server.py -p 8981 --client-num=3" > /ppml/trusted-big-data-ml/secured-argvs
+./init.sh
+SGX=1 ./pal_loader bash 2>&1 | tee fl-server2.log
```
Comment (glorysdj) [inline ppml/trusted-big-data-ml/python/docker-graphene/start-scripts/start-python-fl-server-sgx.sh:9]: fl-server2 -> fl-server-sgx.log

File: ppml/trusted-big-data-ml/python/docker-graphene/start-scripts/start-python-fl-server-sgx.sh [hunk 5]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+cd /ppml/trusted-big-data-ml
+/graphene/Tools/argv_serializer bash -c " /opt/jdk8/bin/java\
+        -cp '/ppml/trusted-big-data-ml/work/spark-3.1.2/conf/:/ppml/trusted-big-data-ml/work/spark-3.1.2/jars/*'\
+        -Xmx10g org.apache.spark.deploy.SparkSubmit\
+        --master 'local[4]'\
+        /ppml/trusted-big-data-ml/work/examples/start-fl-server.py -p 8981 --client-num=3" > /ppml/trusted-big-data-ml/secured-argvs
```
Comment (glorysdj) [inline ppml/trusted-big-data-ml/python/docker-graphene/start-scripts/start-python-fl-server-sgx.sh:7]: is client-num determined when server starts?

Comment (glorysdj) [reply ppml/trusted-big-data-ml/python/docker-graphene/start-scripts/start-python-fl-server-sgx.sh:7]: better to pass args for port and client-num

Comment (Litchilitchy) [reply ppml/trusted-big-data-ml/python/docker-graphene/start-scripts/start-python-fl-server-sgx.sh:7]: > is client-num determined when server starts?

Yes, server must know the client number before starting

File: ppml/trusted-big-data-ml/python/docker-graphene/Dockerfile [hunk 6]
```diff
@@ -165,7 +165,8 @@ RUN apt-get update --fix-missing && \
 RUN wget https://raw.githubusercontent.com/intel-analytics/analytics-zoo/bigdl-2.0/docker/hyperzoo/download-bigdl.sh && \
     chmod a+x ./download-bigdl.sh
 RUN ./download-bigdl.sh && \
-    rm bigdl*.zip
+    rm bigdl*.zip && \
+    cp -r ${BIGDL_HOME}/python/start-fl-server.py /ppml/trusted-big-data-ml/fl
```
Comment (glorysdj) [inline ppml/trusted-big-data-ml/python/docker-graphene/Dockerfile:169]: -r is not needed since start-fl-server.py is a file

**Review by glorysdj — APPROVED** [review]
LGTM

**Pull request merged**